lib/dialects/abstract/connection-manager.js

```diff
@@ -265,31 +265,23 @@ class ConnectionManager {
       promise = Promise.resolve();
     }
 
-    return promise.then(() =>
-      new Promise((resolve, reject) => {
-        const connectionPromise = this.pool.acquire(options.priority, options.type, options.useMaster);
-        const connectionTimer = timers.setInterval(() => {
-          let evictTimer = false;
-
-          if (connectionPromise.isFulfilled()) {
-            resolve(connectionPromise);
-            debug('connection acquire');
-            evictTimer = true;
-          } else if (this.poolError) {
+    return promise.then(() => {
+      return Promise.race([
+        this.pool.acquire(options.priority, options.type, options.useMaster),
+        new Promise((resolve, reject) =>
+        timers.setTimeout(() => {
+          if (this.poolError) {
             reject(this.poolError);
-            this.poolError = null;
-            evictTimer = true;
-          } else if (connectionPromise.isRejected()) {
-            connectionPromise.catch(reject);
-            evictTimer = true;
           }
-
-          if (evictTimer) {
-            timers.clearInterval(connectionTimer);
-          }
-        }, 0);
-      })
-    );
+        }, 0))
```

Comment:
If it is emitted as an event, then this shouldn't check a property once at the next tick, but instead attach a listener to that event:
```ts
let errorListener;
Promise.race([
  aquire(...),
  new Promise((resolve, reject) => {
    errorListener = reject;
    connection.on('error', errorListener);
  })
]).finally(() => {
  connection.removeListener('error', errorListener)
})
```
Only checking the error property on the next tick, once, is brittle because if the error event happens to be emitted on a later tick then it is not caught.